library/src/main/java/com/datatorrent/lib/join/AbstractManagedStateInnerJoinOperator.java

```diff
@@ -0,0 +1,265 @@
+  private int noOfBuckets = 1;
+  private Long bucketSpanTime;
+  protected ManagedTimeStateImpl stream1Store;
+  protected ManagedTimeStateImpl stream2Store;
+
+  /**
+   * Create Managed states and stores for both the streams.
+   */
+  @Override
+  public void createStores()
+  {
+    stream1Store = new ManagedTimeStateImpl();
+    stream2Store = new ManagedTimeStateImpl();
+    stream1Store.setNumBuckets(noOfBuckets);
+    stream2Store.setNumBuckets(noOfBuckets);
+    if (bucketSpanTime != null) {
+      stream1Store.getTimeBucketAssigner().setBucketSpan(Duration.millis(bucketSpanTime));
+      stream2Store.getTimeBucketAssigner().setBucketSpan(Duration.millis(bucketSpanTime));
+    }
+
+    if (getStream1ExpiryTime() != null) {
+      stream1Store.getTimeBucketAssigner().setExpireBefore(Duration.millis(getStream1ExpiryTime()));
+    }
+    if (getStream2ExpiryTime() != null) {
+      stream2Store.getTimeBucketAssigner().setExpireBefore(Duration.millis(getStream2ExpiryTime()));
+    }
+
+    component = new ManagedSpillableComplexComponent();
+    stream1Data = ((ManagedSpillableComplexComponent)component).newSpillableByteArrayListMultimap(stream1Store, !isStream1KeyPrimary());
+    stream2Data = ((ManagedSpillableComplexComponent)component).newSpillableByteArrayListMultimap(stream2Store, !isStream2KeyPrimary());
+  }
+
+  /**
+   * Process the tuple which are received from input ports with the following steps:
+   * 1) Extract key from the given tuple
+   * 2) Insert <key,tuple> into the store where store is the stream1Data if the tuple
+   * receives from stream1 or viceversa.
+   * 3) Get the values of the key in asynchronous if found it in opposite store
+   * 4) If the future is done then Merge the given tuple and values found from step (3) otherwise
+   *    put it in waitingEvents
+   * @param tuple given tuple
+   * @param isStream1Data Specifies whether the given tuple belongs to stream1 or not.
+   */
+  @Override
+  protected void processTuple(T tuple, boolean isStream1Data)
+  {
+    Spillable.SpillableByteArrayListMultimap<K,T> store = isStream1Data ? stream1Data : stream2Data;
+    K key = extractKey(tuple,isStream1Data);
+    long timeBucket = extractTime(tuple,isStream1Data);
+    if (!((ManagedTimeStateMultiValue)store).put(key, tuple,timeBucket)) {
+      return;
+    }
+    Spillable.SpillableByteArrayListMultimap<K, T> valuestore = isStream1Data ? stream2Data : stream1Data;
+    Future<List> future = ((ManagedTimeStateMultiValue)valuestore).getAsync(key);
+    if (future.isDone()) {
+      try {
+        joinStream(tuple,isStream1Data, future.get());
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      } catch (ExecutionException e) {
+        throw new RuntimeException(e);
+      }
+    } else {
+      waitingEvents.put(new JoinEvent<>(key,tuple,isStream1Data),future);
+    }
+  }
+
+  @Override
+  public void handleIdleTime()
+  {
+    if (waitingEvents.size() > 0) {
```

Comment:
Can we process only one element in one handleIdleTime callback?
This is to make sure that processing inside handleIdleTime does not block windows.
handleIdleTime is called over and over again anyway.